src/rpc/util.cpp

```diff
@@ -680,6 +670,44 @@ UniValue RPCHelpMan::GetArgMap() const
     return arr;
 }
 
+void RPCArg::MatchesType(const UniValue& request) const
+{
+    if(m_skip_type_check)return;
+    if (IsOptional() && request.isNull()) return;
+    switch (m_type) {
+    case Type::STR_HEX:
+    case Type::STR: {
+        RPCTypeCheckArgument(request, UniValue::VSTR);
+        return;
+    }
+    case Type::NUM: {
+        RPCTypeCheckArgument(request, UniValue::VNUM);
+        return;
+    }
+    case Type::AMOUNT: {
+        // VNUM or VSTR, checked inside AmountFromValue()
+        return;
+    }
+    case Type::RANGE: {
+        // VNUM or VARR, checked inside ParseRange()
+        return;
+    }
+    case Type::BOOL: {
+        RPCTypeCheckArgument(request, UniValue::VBOOL);
+        return;
+    }
+    case Type::OBJ:
+    case Type::OBJ_USER_KEYS: {
+        RPCTypeCheckArgument(request, UniValue::VOBJ);
+        return;
+    }
+    case Type::ARR: {
+        RPCTypeCheckArgument(request, UniValue::VARR);
+        return;
+    }
```

Comment:
What about doing something like this:
```c++
void RPCArg::MatchesType(const UniValue& request) const
{
    if(m_skip_type_check) return;
    if (IsOptional() && request.isNull()) return;
    // Special cases
    if (m_type == Type::AMOUNT || m_type == Type::RANGE) {
        // Type::AMOUNT: VNUM or VSTR, checked inside AmountFromValue()
        // Type::RANGE: VNUM or VARR, checked inside ParseRange()
        return;
    }
    // Check and throw exception if type is invalid
    UniValue::VType type = ToUnivalueType(m_type);
    request.checkType(type);
}
```

So we define the univalue type error string only once (inside `Univalue::checkType`) and not twice (`Univalue ::checkType` and `RPCTypeCheckArgument`).
Which will let us not have to worry anymore about stuff like #26069 and remove `RPCTypeCheckArgument`.

Reply:
If we decide in the future to also check for hex strings, that wouldn't work, because json (and univalue) only knowns strings, not hex string.